lib/scholar/neighbors/brute_knn.ex

```diff
@@ -0,0 +1,230 @@
+defmodule Scholar.Neighbors.BruteKNN do
+  @moduledoc """
+  Brute-Force k-Nearest Neighbor Search Algorithm.
+
+  In order to find the k-nearest neighbors the algorithm calculates
+  the distance between the query point and each of the data samples.
+  Therefore, its time complexity is $O(MN)$ for $N$ samples and $M$ query points.
+  It uses $O(BN)$ memory for batch size $B$.
+  Larger batch sizes will lead to faster predictions, but will consume more memory.
+  """
+  import Nx.Defn
+  import Scholar.Shared
+  require Nx
+
+  @derive {Nx.Container, keep: [:num_neighbors, :metric, :batch_size], containers: [:data]}
+  defstruct [:num_neighbors, :metric, :data, :batch_size]
+
+  opts = [
+    num_neighbors: [
+      required: true,
+      type: :pos_integer,
+      doc: "The number of nearest neighbors."
+    ],
+    metric: [
+      type: {:or, [{:custom, Scholar.Options, :metric, []}, {:fun, 2}]},
+      default: {:minkowski, 2},
+      doc: ~S"""
+      The function that measures distance between two points. Possible values:
+
+      * `{:minkowski, p}` - Minkowski metric. By changing value of `p` parameter (a positive number or `:infinity`)
+      we can set Manhattan (`1`), Euclidean (`2`), Chebyshev (`:infinity`), or any arbitrary $L_p$ metric.
+
+      * `:cosine` - Cosine metric.
+
+      * Anonymous function of arity 2 that takes two rank-1 tensors of same dimension and returns a scalar.
+      """
+    ],
+    batch_size: [
+      type: :pos_integer,
+      doc: "The number of samples in a batch."
+    ]
+  ]
+
+  @opts_schema NimbleOptions.new!(opts)
+
+  @doc """
+  Fits a brute-force k-NN model.
+
+  ## Options
+
+  #{NimbleOptions.docs(@opts_schema)}
+
+  ## Examples
+
+      iex> data = Nx.tensor([[1, 2], [2, 3], [3, 4], [4, 5], [5, 6]])
+      iex> model = Scholar.Neighbors.BruteKNN.fit(data, num_neighbors: 2)
+      iex> model.num_neighbors
+      2
+      iex> model.data
+      #Nx.Tensor<
+        s64[5][2]
+        [
+          [1, 2],
+          [2, 3],
+          [3, 4],
+          [4, 5],
+          [5, 6]
+        ]
+      >
+  """
+  deftransform fit(data, opts) do
+    if Nx.rank(data) != 2 do
+      raise ArgumentError,
+            "expected input tensor to have shape {n_samples, n_features},
+             got tensor with shape: #{inspect(Nx.shape(data))}"
+    end
+
+    opts = NimbleOptions.validate!(opts, @opts_schema)
+
+    metric =
+      case opts[:metric] do
```

Comment:
Following up above: I would move this normalization to Scholar.Options.metric then. This way everywhere can rely on it being a 2-arity function.

Reply:
You mean instead of returning the atom to return the anonymous function of arity 2?
I agree, but some other modules need to be edited as well. For example k-d tree:
https://github.com/elixir-nx/scholar/blob/0d1bcc106b8a1486d08eb67faba2f4326c3d5a7e/lib/scholar/neighbors/kd_tree.ex#L33-L43
And here is how it is used there:
https://github.com/elixir-nx/scholar/blob/0d1bcc106b8a1486d08eb67faba2f4326c3d5a7e/lib/scholar/neighbors/kd_tree.ex#L283-L286
I would honestly prefer metric to be stored as a function inside a field.

Reply:
The main problem is that different algorithms support different metrics. Brute-force search works with literally any metric, while current implementation of random projection forest works only with Euclidean distance. I am not sure about k-d tree; I think it works only with the three metrics specified in the docs.

Reply:
It is fine then to keep this logic only in this module then :) IMO

Reply:
Agreed. I would still suggest converting atoms to functions and storing them as fields. Seems cleaner and doesn't require choosing the right function to call each time `fit` is used.
Also, it might be worth considering removing `Scholar.Options.metric` given the differences in metrics supported between modules.